### PR TITLE
Build: Initial port to 1.21

### PR DESCRIPTION
**Pull request by camnwalter**
Still needs some verifying as well as a whole new Component API, but the game runs.

### CODE REVIEW

File: src/main/kotlin/com/chattriggers/ctjs/api/client/Sound.kt [hunk 1]
```diff
@@ -523,9 +522,12 @@ class Sound(private val config: NativeObject) {
             throw NotImplementedError()
         }
 
-        override fun getName(): String {
-            return "CTJS_Sounds"
-        }
+        override fun getInfo() = ResourcePackInfo(
+            "chattriggers",
```
Comment (mattco98) [inline src/main/kotlin/com/chattriggers/ctjs/api/client/Sound.kt]: Let's switch all the modids to `ctjs` (and make it a const on `CTJS` so we only have it in one place)

File: src/main/kotlin/com/chattriggers/ctjs/api/commands/DynamicCommands.kt [hunk 2]
```diff
@@ -841,13 +842,13 @@ object DynamicCommands : CommandCollection() {
...
 
-        override fun test(item: Item) = impl.test(item.toMC())
+        override fun test(item: Item) = ItemStack.areItemsAndComponentsEqual(itemType.asItem().toMC(), item.toMC())
 
         fun test(type: ItemType) = itemType.getRegistryName() == type.getRegistryName()
     }
 
     data class MessageFormatArgumentWrapper(private val impl: MessageArgumentType.MessageFormat) {
-        val text: String by impl::contents
+        var text = impl.contents
```
Comment (mattco98) [inline src/main/kotlin/com/chattriggers/ctjs/api/commands/DynamicCommands.kt:851]: What's the reason for this change?

File: src/main/kotlin/com/chattriggers/ctjs/api/inventory/Item.kt [hunk 3]
```diff
@@ -83,8 +85,11 @@ class Item(override val mcValue: ItemStack) : CTWrapper<ItemStack> {
     @JvmOverloads
     fun getLore(advanced: Boolean = false): List<TextComponent> {
         mcValue.asMixin<Skippable>().ctjs_setShouldSkip(true)
-        val tooltip = mcValue.getTooltip(Player.toMC(), if (advanced) TooltipContext.ADVANCED else TooltipContext.BASIC)
-            .mapTo(mutableListOf()) { TextComponent(it) }
+        val tooltip = mcValue.getTooltip(TooltipContext.DEFAULT,
```
Comment (mattco98) [inline src/main/kotlin/com/chattriggers/ctjs/api/inventory/Item.kt:88]: Newline before the first argument

File: src/main/kotlin/com/chattriggers/ctjs/api/render/Renderer.kt [hunk 4]
```diff
@@ -738,12 +738,16 @@ object Renderer {
...
         matrixPushCounter = 0
-        pushMatrix(UMatrixStack(stack))
+        if (stack != null)
+            pushMatrix(UMatrixStack(stack))
+
         block()
-        popMatrix()
+
+        if (stack != null)
+            popMatrix()
```
Comment (mattco98) [inline src/main/kotlin/com/chattriggers/ctjs/api/render/Renderer.kt:750]: Not super related to your changes, but I think `block` can execute user-code and thus throw exception, so the push-pop should probably be in a `try`/`finally` block. Might as well do it while you're here

File: src/main/kotlin/com/chattriggers/ctjs/api/world/Server.kt [hunk 5]
```diff
@@ -53,7 +53,6 @@ object Server {
         return toMC()?.label?.let { TextComponent(it) }?.formattedText ?: ""
     }
 
-    // TODO(breaking): Return -1 if not in a world
```
Comment (mattco98) [inline src/main/kotlin/com/chattriggers/ctjs/api/world/Server.kt:56]: Isn't this how it's always been? It's in the ktdoc already

File: src/main/kotlin/com/chattriggers/ctjs/internal/launch/DynamicMixinManager.kt [hunk 6]
```diff
@@ -94,7 +95,7 @@ internal object DynamicMixinManager {
     private object ByteBasedStreamHandler : URLStreamHandler() {
         private val classBytes = mutableMapOf<String, ByteArray>()
 
-        val url = URL(GENERATED_PROTOCOL, null, -1, "/", ByteBasedStreamHandler)
+        val url: URL = URL.of(URI(GENERATED_PROTOCOL, null, "/", ""), ByteBasedStreamHandler)
```
Comment (mattco98) [inline src/main/kotlin/com/chattriggers/ctjs/internal/launch/DynamicMixinManager.kt:98]: Remove the type annotation

File: src/main/kotlin/com/chattriggers/ctjs/internal/commands/CTCommand.kt [hunk 7]
```diff
@@ -260,4 +269,27 @@ internal object CTCommand : Initializer {
             fun getFile(ctx: CommandContext<*>, name: String) = ctx.getArgument(name, File::class.java)
         }
     }
+
+    private object ModuleArgumentType : ArgumentType<String> {
```
Comment (mattco98) [inline src/main/kotlin/com/chattriggers/ctjs/internal/commands/CTCommand.kt:273]: Very nice :)

File: src/main/kotlin/com/chattriggers/ctjs/api/client/Sound.kt [hunk 8]
```diff
@@ -260,6 +258,8 @@ class Sound(private val config: NativeObject) {
      */
     @JvmOverloads
     fun play(delay: Int = 0) {
+        require (World.isLoaded()) { "Can not play a custom sound outside the world" }
```
Comment (mattco98) [inline src/main/kotlin/com/chattriggers/ctjs/api/client/Sound.kt:261]: Hmm, this is a surprising restriction from a user perspective. Let's at least mention it in the class's ksdoc along with a TODO

**Review by mattco98 — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/chattriggers/ctjs/api/commands/DynamicCommands.kt (hunk 2 above)
Comment (camnwalter) [reply src/main/kotlin/com/chattriggers/ctjs/api/commands/DynamicCommands.kt:851]: We can't delegate the property since `MessageArgumentType.MessageFormat` is a Java Record

File: src/main/kotlin/com/chattriggers/ctjs/api/world/Server.kt (hunk 5 above)
Comment (camnwalter) [reply src/main/kotlin/com/chattriggers/ctjs/api/world/Server.kt:56]: It was a breaking change that wasn't documented in the migration doc, so I just added it there. On 2.2, we used to always return 5, which is what we return for singleplayer.

File: src/main/kotlin/com/chattriggers/ctjs/api/client/Sound.kt (hunk 8 above)
Comment (camnwalter) [reply src/main/kotlin/com/chattriggers/ctjs/api/client/Sound.kt:261]: Yeah, maybe in the future someone can look into getting sounds to work, but I couldn't so this is a band-aid fix. Will do.

File: src/main/kotlin/com/chattriggers/ctjs/api/commands/DynamicCommands.kt (hunk 2 above)
Comment (mattco98) [reply src/main/kotlin/com/chattriggers/ctjs/api/commands/DynamicCommands.kt:851]: Gotcha. Does it need to be a `var` though?

Comment (camnwalter) [reply src/main/kotlin/com/chattriggers/ctjs/api/commands/DynamicCommands.kt:851]: Yes since you couldn't set the text otherwise. Honestly I don't know much about these argument wrappers, I just saw that it used delegation to be able to get and set, so I used a var to mimic that.

Comment (mattco98) [reply src/main/kotlin/com/chattriggers/ctjs/api/commands/DynamicCommands.kt:851]: Oh I didn't realize these classes were public API, nvm then

File: src/main/kotlin/com/chattriggers/ctjs/api/world/Server.kt (hunk 5 above)
Comment (mattco98) [reply src/main/kotlin/com/chattriggers/ctjs/api/world/Server.kt:56]: I'm stupid lol I thought you added this comment, not removed it
